File: new-resource-detector/src/test/java/nl/haploid/octowight/service/DirtyResourceProducerServiceTest.java
package nl.haploid.octowight.service;

import mockit.Injectable;
import mockit.Mocked;
import mockit.StrictExpectations;
import mockit.Tested;
import nl.haploid.octowight.JsonMapper;
import nl.haploid.octowight.TestData;
import nl.haploid.octowight.data.Resource;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.junit.Test;

import java.util.concurrent.Future;

import static org.junit.Assert.assertEquals;

public class DirtyResourceProducerServiceTest {

	@Tested
	private DirtyResourceProducerService service;

	@Injectable
	private KafkaProducer<String, String> kafkaProducer;

	@Injectable
	private JsonMapper jsonMapper;

	@Test
	@SuppressWarnings("unchecked")
	public void testSendDirtyResource(final @Mocked Future<RecordMetadata> expectedFuture) throws Exception {
		final Resource resource = TestData.resource(451l);
		final String message = "joy";
		new StrictExpectations() {{
			jsonMapper.toString(resource);
			times = 1;
			result = message;
			kafkaProducer.send((ProducerRecord<String, String>) any);
			times = 1;
			result = expectedFuture;
		}};
		final Future<RecordMetadata> actualFuture = service.sendDirtyResource(resource);
		assertEquals(expectedFuture, actualFuture);
	}

	@Test
	public void testResolveFuture(final @Mocked Future<RecordMetadata> future,
								  final @Mocked RecordMetadata expectedMetadata) throws Exception {
		new StrictExpectations() {{
			future.get();
			times = 1;
			result = expectedMetadata;
		}};
		final RecordMetadata actualMetadata = service.resolveFuture(future);
		assertEquals(expectedMetadata, actualMetadata);
	}
}

File: event-consumer/src/test/scala/nl/haploid/octowight/service/DirtyResourceProducerServiceTest.scala
package nl.haploid.octowight.service

import java.util.concurrent.Future

import nl.haploid.octowight._
import nl.haploid.octowight.kafka.KafkaProducerFactory
import nl.haploid.octowight.registry.data.ResourceIdentifier
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord, RecordMetadata}
import org.easymock.EasyMock
import org.springframework.test.util.ReflectionTestUtils

class DirtyResourceProducerServiceTest extends AbstractTest {
  @Tested private[this] val dirtyResourceProducerService: DirtyResourceProducerService = null
  @Mocked private[this] val kafkaProducerFactory: KafkaProducerFactory = null
  @Mocked private[this] val jsonMapper: JsonMapper = null

  override def beforeEach() = {
    super.beforeEach()
    ReflectionTestUtils.setField(dirtyResourceProducerService, "topic", TestData.topic)
  }

  "Dirty resource producer" should "send dirty resource" in {
    val kafkaProducer = mock[KafkaProducer[String, String]]
    val expectedFuture = mock[Future[RecordMetadata]]
    val resourceRoot = TestData.resourceRoot(TestData.nextLong)
    val resourceIdentifier = new ResourceIdentifier(collection = resourceRoot.resourceCollection, id = resourceRoot.resourceId)
    val message = "joy"
    expecting {
      kafkaProducerFactory.kafkaProducer andReturn kafkaProducer once()
      jsonMapper.serialize(resourceIdentifier) andReturn message once()
      kafkaProducer.send(EasyMock.anyObject[ProducerRecord[String, String]]) andReturn expectedFuture once()
    }
    whenExecuting(jsonMapper, kafkaProducerFactory, kafkaProducer) {
      val actualFuture = dirtyResourceProducerService.sendDirtyResource(resourceRoot)
      actualFuture should be(expectedFuture)
    }
  }
}
